asim: add zone config satisfiability check

wenyihu6 opened the pull request:
**asim: correct a typo in ComplexConfig**

This patch corrects a typo where "US_East_3" was duplicated. It corrects it to 
"US_East_4".

Release Note: none
Epic: none

----

**asim: add zone config satisfiability check**

Now that we have added the option to generate random span configurations in
#110967, we want to have a way to check whether these configurations are
satisfiable with the cluster setting.

This patch adds the validation check. Please note that the validation process can
be expensive with a time complexity of O(max(node count in the cluster, number of
replica constraints, number of voter constraints)). To perform this validation
and see which span config could lead to failure, please use following command:

```
"eval" [verbose=validate]
```

See also: https://github.com/cockroachdb/cockroach/pull/110967
Part of: https://github.com/cockroachdb/cockroach/issues/106192
Release Note: none
Epic: none

wenyihu6 committed:
asim: add zone config satisfiability check

Now that we have added the option to generate random span configurations in
#110967, we want to have a way to check whether these configurations are
satisfiable with the cluster setting.

This patch adds the validation check. Please note that the validation process can
be expensive with a time complexity of O(max(node count in the cluster, number of
replica constraints, number of voter constraints)). To perform this validation
and see which span config could lead to failure, please use following command:

```
"eval" [verbose=validate]
```

See also: #110967
Part of: #106192
Release Note: none
Epic: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/110968)
<!-- Reviewable:end -->

wenyihu6 commented:
Please only review the last commit. The preceding commits are from #110967.

wenyihu6 committed:
asim: correct a typo in ComplexConfig

This patch corrects a typo where "US_East_3" was duplicated. It corrects it to
"US_East_4".

Release Note: none
Epic: none

wenyihu6 commented:
I expect that the second commit will require a few iterations of review. Let me know if you would like me to pull the first commit into its own PR to get the trivial part in first

kvoli reviewed (commented):
I'm still reviewing the `config_validator.go` changes. Flushing some comments.

> Reviewed 5 of 5 files at r1, 9 of 14 files at r2, 3 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110968#-:-NfGxttl4J7P_wFO_-68:b94uki7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/asim/validator/config_validator.go` line 156 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110968#-NfGs8S84v0ZnWwkkx0v:-NfGs8S84v0ZnWwkkx0w:b-aogs81) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a59738bfaed494a1dbaad51cfd082dbefc151f/pkg/kv/kvserver/asim/validator/config_validator.go#L156)):*
> ```Go
> // Otherwise, it returns error (not `nil`).
> func (m *mockAllocator) processConstraintsHelper(
> 	constraintsConjunction []roachpb.ConstraintsConjunction,
> ```

nit: naming this `constraintConjunctions` would be clearer, because this is a bag of conjunctions.
___
*[`pkg/kv/kvserver/asim/validator/config_validator.go` line 384 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110968#-NfGycW_0x1igrbaDdMT:-NfGycWaARTWG00Eovte:b-qjd7zq) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a59738bfaed494a1dbaad51cfd082dbefc151f/pkg/kv/kvserver/asim/validator/config_validator.go#L384)):*
> ```Go
> // non-voters are then assigned as needed. If any zones or regions lack
> // available nodes for assignment, the constraint is considered as
> // unsatisfiable.
> ```

This is a great high level summary of the algorithm. Could you expand this further, and explain why this algorithm will work given the limitations (assumptions)?

IIUC the algorithm works because of the structure of region-zone hierarchy. Where there intersection of any two region-region or zone-zone combos is empty, and node's only have a region and zone locality. 

Could you think of an example where this wouldn't work, if we allow for intersecting constraint conjunctions? No need to solve this here, but it would be good to write out.
___
*[`pkg/kv/kvserver/asim/validator/validator_test.go` line 172 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110968#-NfGq_7xD2n6Sp1e9Mc6:-NfGq_7xD2n6Sp1e9Mc7:b-dj3t3x) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a59738bfaed494a1dbaad51cfd082dbefc151f/pkg/kv/kvserver/asim/validator/validator_test.go#L172)):*
> ```Go
> 			description: "insufficient replicas for region constraint",
> 			constraint: "num_replicas=28 num_voters=28 " +
> 				"constraints={'+region=US_East':17,'+region=US_West':2,'+region=EU':10}",
> ```

nit: wouldn't this also fail due to sum(constraints) not matching the num_replicas (28 vs 29)? It might help to reduce `EU:9` to make the test case focus on just `US_East` having a constrained num_replicas greater than the stores in number of stores in `US_East`.
___
*[`pkg/kv/kvserver/asim/validator/validator_test.go` line 233 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110968#-NfGramu3DqEv5CQCdY6:-NfGramu3DqEv5CQCdY7:b10ca2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a59738bfaed494a1dbaad51cfd082dbefc151f/pkg/kv/kvserver/asim/validator/validator_test.go#L233)):*
> ```Go
> 		},
> 	}
> 	for _, tc := range testCases {
> ```

Nice tests!


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/kv/kvserver/asim/validator/config_validator.go` line 384 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110968#-NfGycW_0x1igrbaDdMT:-NfH8Z3i230hIWstd_6S:bgp2q0s) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a59738bfaed494a1dbaad51cfd082dbefc151f/pkg/kv/kvserver/asim/validator/config_validator.go#L384)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

I'm not sure if I'm understanding this correctly. Could you give me an example on what non-empty region-region combo means and when node has more than one region or zone locality? My understanding was that regions or zones are non-overlapping. And each node have unique region and zone locality. And [clusterInfo.Regions](https://github.com/cockroachdb/cockroach/blob/18094543f560aea3195e5823ca2e9b4812cce4a0/pkg/kv/kvserver/asim/state/config_loader.go#L264-L267) would specify regions, zones, and nodes under it.
</blockquote></details>

or do you mean there exists duplicate constraints for the same region value (something like constraints={'+region=US_East':1,'+region=US_East':2})?


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110968)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/asim/validator/config_validator.go` line 384 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110968#-NfGycW_0x1igrbaDdMT:-NfHB-4VBRpnx3TK2Pps:bidomz4) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a59738bfaed494a1dbaad51cfd082dbefc151f/pkg/kv/kvserver/asim/validator/config_validator.go#L384)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

or do you mean there exists duplicate constraints for the same region value (something like constraints={'+region=US_East':1,'+region=US_East':2})?
</blockquote></details>

Using attributes, which constraints can match on, it is possible, and something we support (begrudgingly at this stage).

e.g. `locality=region=a,zone=a1 attributes=disk` 


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
a

wenyihu6 committed:
a

wenyihu6 committed:
a

wenyihu6 committed:
a

wenyihu6 committed:
a
